### PR TITLE
pkg/ip: add functionality to coalesce CIDR list

### PR DESCRIPTION
**Pull request by ianvernon**
Add CoalesceCIDRs and associated helper functions, which combines a list of
CIDRs into the smallest equivalent set of CIDRs.

Signed-off by: Ian Vernon <ian@cilium.io>

This was long-lived in a previous PR, but I wanted to separate the library code being merged from the actual consumption of the code in Cilium.

Partially address #1658 : this is only the library functionality. Cilium can consume this change at a later time (this is tracked as part of #1658).

### CODE REVIEW

**Comment by ianvernon** [comment]
test-me-please

File: pkg/ip/ip.go [hunk 1]
```diff
@@ -41,6 +42,15 @@ func CountIPsInCIDR(cidr string) int {
 	return 1<<uint(size-subnet) - 1
 }
 
+var (
+	v4Asv6            = []byte{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xff, 0xff}
```
Comment (joestringer) [inline pkg/ip/ip.go]: I assume this is making use of the RFC2765 IPv4-mapped address prefix? Would be nice to have a brief comment mentioning this, and perhaps naming the variable as `v4Mappedv6Prefix`

**Review by joestringer — APPROVED** [review]
Minor comment, nothing blocking. I expect to start using this library soon, after which point I may be able to provide some better feedback as a user. Either way, the testing looks good and I expect that there will be users in the codebase before too long so it seems reasonable to merge in its current state.

File: pkg/ip/ip.go (hunk 1 above)
Comment (ianvernon) [reply pkg/ip/ip.go]: Fixed in latest update.

**Comment by ianvernon** [comment]
test-me-please

**Pull request merged**